src/lattice_reaction_system_diffusion.jl

```diff
@@ -0,0 +1,461 @@
+        new(Num(rate), species, ModelingToolkit.getname(species))
+    end
+    function DiffusionReaction(rate::Symbol, species::Num)
+        new(Symbolics.variable(rate), species, ModelingToolkit.getname(species))
+    end
+    function DiffusionReaction(rate::Num, species::Symbol)
+        new(rate, Symbolics.variable(species), species)
+    end
+    function DiffusionReaction(rate::Number, species::Symbol)
+        new(Num(rate), Symbolics.variable(species), species)
+    end
+    function DiffusionReaction(rate::Symbol, species::Symbol)
+        new(Symbolics.variable(rate), Symbolics.variable(species), species)
+    end
+end
+# Creates a vector of DiffusionReactions.
+function diffusion_reactions(diffusion_reactions)
+    [DiffusionReaction(dr[1], dr[2]) for dr in diffusion_reactions]
+end
+# Gets the parameters in a diffusion reaction.
+ModelingToolkit.parameters(dr::DiffusionReaction) = Symbolics.get_variables(dr.rate)
+
+### Lattice Reaction Network Structure ###
+# Desribes a spatial reaction network over a graph.
+struct LatticeReactionSystem # <: MT.AbstractTimeDependentSystem # Adding this part messes up show, disabling me from creating LRSs
+    """The reaction system within each comaprtment."""
+    rs::ReactionSystem
+    """The spatial reactions defined between individual nodes."""
+    spatial_reactions::Vector{<:AbstractSpatialReaction}
+    """The graph on which the lattice is defined."""
+    lattice::DiGraph
+
+    # Derrived values.
+    """The number of compartments."""
+    nC::Int64
+    """The number of edges."""
+    nE::Int64
+    """The number of species."""
+    nS::Int64
+    """Whenever the initial input was a di graph."""
+    init_digraph::Bool
+
+    function LatticeReactionSystem(rs::ReactionSystem,
+                                   spatial_reactions::Vector{<:AbstractSpatialReaction},
+                                   lattice::DiGraph; init_digraph = true)
+        return new(rs, spatial_reactions, lattice, nv(lattice), ne(lattice),
+                   length(species(rs)), init_digraph)
+    end
+    function LatticeReactionSystem(rs::ReactionSystem,
+                                   spatial_reactions::Vector{<:AbstractSpatialReaction},
+                                   lattice::SimpleGraph)
+        return LatticeReactionSystem(rs, spatial_reactions, graph_to_digraph(lattice);
+                                     init_digraph = false)
+    end
+    function LatticeReactionSystem(rs::ReactionSystem,
+                                   spatial_reaction::AbstractSpatialReaction,
+                                   lattice::Graphs.AbstractGraph)
+        return LatticeReactionSystem(rs, [spatial_reaction], lattice)
+    end
+end
+# Covnerts a graph to a digraph (in a way where we know where the new edges are in teh edge vector).
+function graph_to_digraph(g1)
+    g2 = Graphs.SimpleDiGraphFromIterator(reshape(permutedims(hcat(collect(edges(g1)),
```

Comment:
This seems like it is likely excessively allocating / copying the underlying data multiple times. For a large graph, as would be common in PDE discretizations, that will be pretty memory wasteful.

Reply:
I will think if there is a better way to do it.

Reply:
Can you not just manually loop through and construct the `DiGraph`?

Reply:
There was something where I wanted to ensure that things had a specific order in the new graph. But I might also have gotten overzealous in trying to write it as a single compact statement as well.

Reply:
To be honest, I have no clue what this one line is doing. So just from a readability standpoint it might be better to expand this out a bit.

Reply:
I think you are probably right

Reply:
I'm just imagining making updates to this code in two years and wondering what the heck is going on here...

Reply:
Currently working on this. There are some complications with how DiGraphs are iterated through (compared to Graphs). Best case I will get to this this weekend.

Reply:
I have now redone this part.

Since Graphs.jl always iterates through edges in the order of their source vertex index (followed by the destination source index), trying to create a new DiGraph with tactically ordered edges was a poor idea.

Instead, I now simply run `DiGraph(lattice)`, and then have a special function for mapping edge parameters to their correct portion.

This is all only for the special case where someone gives a non-directed graph of size n as input, which is then internally converted to a directed graph of size 2n. Then, when they provide values for one (or more) parameters of length n (indicating different values for each edge) we need to convert this correctly to a vector of length 2n (assuming the same values in both direction for each edge).